src/common/column_matrix.h

```diff
@@ -103,96 +110,179 @@ class ColumnMatrix {
+    } else {
+      missing_flags_.resize(boundary_[nfeature - 1].index_end, true);
+    }
+
+    if (all_dense) {
+      switch (gmat.index.getBinBound()) {
+        case POWER_OF_TWO_8:
+          SetIndexAllDense(gmat.index.data<uint8_t>(), gmat, nrow, nfeature, noMissingValues);
+          break;
+        case POWER_OF_TWO_16:
+          SetIndexAllDense(gmat.index.data<uint16_t>(), gmat, nrow, nfeature, noMissingValues);
+          break;
+        case POWER_OF_TWO_32:
+          SetIndexAllDense(gmat.index.data<uint32_t>(), gmat, nrow, nfeature, noMissingValues);
+          break;
+      }
+    } else {
+      switch (type_size_) {
+        case 1:
+          SetIndex<uint8_t>(gmat.index.data<uint32_t>(), gmat, nrow, nfeature);
+          break;
+        case 2:
+          SetIndex<uint16_t>(gmat.index.data<uint32_t>(), gmat, nrow, nfeature);
+          break;
+        case 4:
+          SetIndex<uint32_t>(gmat.index.data<uint32_t>(), gmat, nrow, nfeature);
+          break;
+      }
+    }
+  }
+
+  /* Fetch an individual column. This code should be used with XGBOOST_TYPE_SWITCH
+     to determine type of bin id's */
+  template <typename T>
+  inline Column<T> GetColumn(unsigned fid) const {
+    Column<T> c(type_[fid],
+                reinterpret_cast<const T*>(&index_[boundary_[fid].index_begin * type_size_]),
+                index_base_[fid], (type_[fid] == ColumnType::kSparseColumn ?
+                &row_ind_[boundary_[fid].index_begin] : nullptr),
+                boundary_[fid].index_end - boundary_[fid].index_begin,
+                &missing_flags_, boundary_[fid].index_begin);
+    return c;
+  }
+
+  template<typename T>
+  inline void SetIndexAllDense(T* index, const GHistIndexMatrix& gmat,  const size_t nrow,
+                               const size_t nfeature,  const bool noMissingValues) {
+    T* local_index = reinterpret_cast<T*>(&index_[0]);
+
+    if (noMissingValues) {
+      const int32_t nthread = omp_get_max_threads();
+      #pragma omp parallel for num_threads(nthread)
+      for (omp_ulong rid = 0; rid < nrow; ++rid) {
+        const size_t ibegin = rid*nfeature;
+        const size_t iend = (rid+1)*nfeature;
+        size_t j = 0;
+        for (size_t i = ibegin; i < iend; ++i, ++j) {
+            const size_t idx = boundary_[j].index_begin;
+            local_index[idx + rid] = index[i];
+        }
+      }
+    } else {
+      size_t rbegin = 0;
+      for (const auto &batch : gmat.p_fmat_->GetBatches<SparsePage>()) {
+        for (size_t rid = 0; rid < batch.Size(); ++rid) {
+          SparsePage::Inst inst = batch[rid];
+          const size_t ibegin = gmat.row_ptr[rbegin + rid];
+          const size_t iend = gmat.row_ptr[rbegin + rid + 1];
+          CHECK_EQ(ibegin + inst.size(), iend);
+          size_t j = 0;
+          size_t fid = 0;
+          for (size_t i = ibegin; i < iend; ++i, ++j) {
+              fid = inst[j].index;
+              const size_t idx = boundary_[fid].index_begin;
```

Comment:
What's the logic behind `rbegin + idx`, what is it pointing to?

Reply:
Number of elements in index_, gmat.index is a sum of `SparsePage` batch sizes (nrow). So before this PR, `rid` index was in interval [0, nrow), but now we have to process all `SparsePage` batches.
It's more about `rbegin + rid` : `rbegin` variable is responsible for row offset to handle all batches.
Comment was added in code.

Reply:
comments were added.